rc/src/lib.rs

```diff
@@ -0,0 +1,60 @@
+use std::ops::{Deref,Drop};
+use std::{mem,ptr};
+
+pub struct MyRc<T> {
+    ptr: *mut RcT<T>
+}
+
+struct RcT<T> {
+    data: T,
+    count: usize
+}
+
+impl<T> MyRc<T> {
+
+    pub fn new(t : T) -> MyRc<T> {
+        let data = RcT { data: t, count: 1 };
+        MyRc { ptr: Box::into_raw(Box::new(data)) }
+    }
+
+    pub fn consume(self) -> Result<T,MyRc<T>> {
+        unsafe {
+            if (*self.ptr).count == 1 {
+                let mut t : T = mem::uninitialized();
+                mem::swap(&mut (*self.ptr).data, &mut t);
+                Ok(t)
```

Comment:
At the end of this block `self` goes out of scope and its destructor runs. The destructor will ultimately run the destructor for the uninitialized `T`, which is a problem.

Valgrind picks this up (see the piazza post for how to run valgrind on Rust programs), but the test suite misses this, so no penalty for missing this.